File: mlx_sparse/_csc.py
# Copyright (c) 2026 The mlx-sparse contributors - All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from dataclasses import dataclass

import mlx.core as mx

import mlx_sparse._native as _native
from mlx_sparse._csr import CSRArray
from mlx_sparse._typing import Shape2D, ValidationMode
from mlx_sparse._validation import (
    ensure_mx_array,
    normalize_shape,
    normalize_validation_mode,
    sanitize_scalar,
    validate_csc_metadata,
    validate_csc_values,
)


@dataclass(frozen=True, slots=True)
class CSCArray:
    """A 2D sparse matrix in Compressed Sparse Column (CSC) format.

    CSCArray stores a 2D sparse matrix using three MLX arrays:

    - **data**: non-zero values, shape ``(nnz,)``.
    - **indices**: row index of each stored value, shape ``(nnz,)``.
    - **indptr**: column pointer array, shape ``(n_cols + 1,)``.

    Column ``j`` spans ``data[indptr[j] : indptr[j+1]]`` with corresponding
    row indices ``indices[indptr[j] : indptr[j+1]]``. Duplicate row entries
    within a column are permitted unless the matrix is in canonical form.

    CSC is the column-compressed dual of CSR. It is the natural layout for
    operations that consume one full column at a time, such as transpose
    matvec, column-oriented canonicalization, and future direct factorization
    kernels.

    Args:
        data: Non-zero values, shape ``(nnz,)``.
        indices: Row indices, shape ``(nnz,)``.
        indptr: Column pointer array, shape ``(n_cols + 1,)``.
        shape: Matrix dimensions as ``(n_rows, n_cols)``.
        sorted_indices: Hint that row indices within each column are sorted
            ascending. Defaults to ``False``.
        has_canonical_format: Hint that the matrix has sorted row indices and
            no duplicate row index in any column. Implies
            ``sorted_indices=True``. Defaults to ``False``.
    """

    data: mx.array
    indices: mx.array
    indptr: mx.array
    shape: Shape2D
    sorted_indices: bool = False
    has_canonical_format: bool = False

    def __post_init__(self) -> None:
        object.__setattr__(self, "shape", normalize_shape(self.shape))

    @property
    def nnz(self) -> int:
        """Number of stored values (including any duplicates)."""
        return int(self.data.shape[0])

    @property
    def dtype(self):
        """Value dtype of the stored non-zeros."""
        return self.data.dtype

    @property
    def index_dtype(self):
        """Integer dtype used for ``indices`` and ``indptr``."""
        return self.indices.dtype

    @property
    def ndim(self) -> int:
        """Always 2. Sparse arrays in this package are rank-2."""
        return 2

    def __repr__(self) -> str:
        return (
            "CSCArray("
            f"shape={self.shape}, nnz={self.nnz}, dtype={self.dtype}, "
            f"index_dtype={self.index_dtype}, "
            f"sorted_indices={self.sorted_indices}, "
            f"has_canonical_format={self.has_canonical_format})"
        )

    def todense(self) -> mx.array:
        """Materialize the sparse matrix as a dense MLX array."""
        return _native.csc_todense(self.data, self.indices, self.indptr, self.shape)

    def row_sums(self) -> mx.array:
        """Return the sum of stored values in each CSC row."""
        from mlx_sparse._ops import csc_row_sums

        return csc_row_sums(self)

    def col_sums(self) -> mx.array:
        """Return the sum of stored values in each CSC column."""
        from mlx_sparse._ops import csc_col_sums

        return csc_col_sums(self)

    def column_sums(self) -> mx.array:
        """Alias for :meth:`col_sums`."""
        return self.col_sums()

    def row_norms(self) -> mx.array:
        """Return the dense-semantics L2 norm of each CSC row as ``float32``."""
        from mlx_sparse._ops import csc_row_norms

        return csc_row_norms(self)

    def col_norms(self) -> mx.array:
        """Return the dense-semantics L2 norm of each CSC column as ``float32``."""
        from mlx_sparse._ops import csc_col_norms

        return csc_col_norms(self)

    def column_norms(self) -> mx.array:
        """Alias for :meth:`col_norms`."""
        return self.col_norms()

    def diagonal(self) -> mx.array:
        """Return the summed main diagonal."""
        from mlx_sparse._ops import csc_diagonal

        return csc_diagonal(self)

    def trace(self) -> mx.array:
        """Return the summed main diagonal as a scalar."""
        from mlx_sparse._ops import csc_trace

        return csc_trace(self)

    def sum(self, axis=None) -> mx.array:
        """Sum sparse values over all entries, rows, or columns.

        ``axis=None`` returns a scalar, ``axis=1`` returns row sums, and
        ``axis=0`` returns column sums.
        """
        if axis is None:
            return mx.sum(self.col_sums())
        if axis in (1, -1):
            return self.row_sums()
        if axis in (0, -2):
            return self.col_sums()
        raise ValueError(f"CSCArray.sum axis must be None, 0, or 1; got {axis!r}.")

    def tocsr(self, *, canonical: bool | None = None) -> CSRArray:
        """Convert to :class:`~mlx_sparse.CSRArray`.

        Args:
            canonical: If ``True``, canonicalize the returned CSR matrix. If
                ``False`` or ``None`` (default), return the structural
                conversion as produced by the native count/prefix/fill path.
                The structural path preserves values but does not promise
                sorted output metadata on every backend.
        """
        data, indices, indptr = _native.csc_tocsr(
            self.data,
            self.indices,
            self.indptr,
            self.shape,
        )
        out = CSRArray(
            data=data,
            indices=indices,
            indptr=indptr,
            shape=self.shape,
            sorted_indices=False,
            has_canonical_format=False,
        )
        if canonical is True:
            return out.canonicalize()
        if canonical is False:
            return CSRArray(
                data=out.data,
                indices=out.indices,
                indptr=out.indptr,
                shape=out.shape,
                sorted_indices=out.sorted_indices,
                has_canonical_format=False,
            )
        return out

    def sort_indices(self) -> "CSCArray":
        """Return a new CSCArray with row indices sorted within each column."""
        if self.sorted_indices:
            return self
        data, indices, indptr = _native.csc_sort_indices(
            self.data,
            self.indices,
            self.indptr,
        )
        return CSCArray(
            data=data,
            indices=indices,
            indptr=indptr,
            shape=self.shape,
            sorted_indices=True,
            has_canonical_format=False,
        )

    def sum_duplicates(self) -> "CSCArray":
        """Sum duplicate row entries within each column."""
        sorted_self = self.sort_indices()
        data, indices, indptr = _native.csc_sum_duplicates(
            sorted_self.data,
            sorted_self.indices,
            sorted_self.indptr,
        )
        return CSCArray(
            data=data,
            indices=indices,
            indptr=indptr,
            shape=self.shape,
            sorted_indices=True,
            has_canonical_format=True,
        )

    def canonicalize(self) -> "CSCArray":
        """Return canonical form: sorted row indices, no duplicates."""
        if self.has_canonical_format:
            return self
        return self.sum_duplicates()

    def transpose(self) -> CSRArray:
        """Transpose the sparse matrix, returning a zero-copy CSRArray."""
        return CSRArray(
            data=self.data,
            indices=self.indices,
            indptr=self.indptr,
            shape=(self.shape[1], self.shape[0]),
            sorted_indices=self.sorted_indices,
            has_canonical_format=self.has_canonical_format,
        )

    @property
    def T(self) -> CSRArray:
        """Transposed matrix. Alias for :meth:`transpose`."""
        return self.transpose()

    def conj(self) -> "CSCArray":
        """Complex-conjugate the stored values."""
        return CSCArray(
            data=mx.conjugate(self.data),
            indices=self.indices,
            indptr=self.indptr,
            shape=self.shape,
            sorted_indices=self.sorted_indices,
            has_canonical_format=self.has_canonical_format,
        )

    def conjugate(self) -> "CSCArray":
        """Alias for :meth:`conj`."""
        return self.conj()

    @property
    def H(self) -> CSRArray:
        """Hermitian (conjugate) transpose. Equivalent to ``conj().T``."""
        return self.conj().transpose()

    def __matmul__(self, rhs):
        """Matrix multiplication via the ``@`` operator."""
        from mlx_sparse._coo import COOArray
        from mlx_sparse._ops import csc_matmat, csc_matmul, csc_matvec

        if isinstance(rhs, CSCArray):
            return csc_matmat(self, rhs)
        if isinstance(rhs, CSRArray | COOArray):
            raise NotImplementedError(
                "Mixed-format CSC sparse-sparse matmul is not implemented. "
                "Convert explicitly if another format is acceptable for your workload."
            )

        rhs = ensure_mx_array(rhs)
        if rhs.ndim == 1:
            return csc_matvec(self, rhs)
        if rhs.ndim >= 2:
            return csc_matmul(self, rhs)
        raise ValueError(f"CSC matmul expects rank-1 or higher RHS, got {rhs.shape}.")

    def __rmul__(self, other):
        """Multiply the current CSCArray by a number using the ``*`` operator.

        This returns a new CSCArray with the data multiplied by the number, and
        therefore does not in-place mutate the current CSCArray.

        Args:
            other: A valid number (complex or not).

        Returns:
            A new CSCArray with the data multiplied by the number.

        Raises:
            TypeError: If ``other`` is not an actual number.
        """
        other = sanitize_scalar(other)

        return CSCArray(
            data=other * self.data,
            indices=self.indices,
            indptr=self.indptr,
            shape=self.shape,
            sorted_indices=self.sorted_indices,
            has_canonical_format=self.has_canonical_format,
        )

    def __mul__(self, other):
        """Multiply the current CSCArray by a number using the ``*`` operator.

        This returns a new CSCArray with the data multiplied by the number, and
        therefore does not in-place mutate the current CSCArray.

        Args:
            other: A valid number (complex or not).

        Returns:
            A new CSCArray with the data multiplied by the number.

        Raises:
            TypeError: If ``other`` is not an actual number.
        """
        return self.__rmul__(other)


def csc_array(
    arg,
    shape,
    *,
    validate: ValidationMode = "metadata",
    sorted_indices: bool = False,
    canonical: bool | None = None,
) -> CSCArray:
    """Construct a :class:`CSCArray` from explicit CSC buffers."""
    mode = normalize_validation_mode(validate)
    shape = normalize_shape(shape)

    if isinstance(arg, CSCArray):
        array = arg
        if array.shape != shape:
            raise ValueError(
                f"CSCArray shape mismatch: got {array.shape}, expected {shape}."
            )
        return array

    try:
        data, indices, indptr = arg
    except Exception as exc:
        raise TypeError(
            "csc_array expects (data, indices, indptr) or a CSCArray instance."
        ) from exc

    data = ensure_mx_array(data)
    indices = ensure_mx_array(indices)
    indptr = ensure_mx_array(indptr)

    if mode != "none":
        validate_csc_metadata(data, indices, indptr, shape)
    if mode == "full":
        validate_csc_values(indices, indptr, shape, data.shape[0])

    has_canonical_format = bool(canonical) if canonical is not None else False
    if has_canonical_format:
        sorted_indices = True

    return CSCArray(
        data=data,
        indices=indices,
        indptr=indptr,
        shape=shape,
        sorted_indices=sorted_indices,
        has_canonical_format=has_canonical_format,
    )

File: src/random/csc.cpp
// Copyright (c) 2026 The mlx-sparse contributors - All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "random/random.h"

#include <algorithm>
#include <cstdint>
#include <stdexcept>
#include <tuple>
#include <vector>

#include "common/cpu_parallel.h"
#include "mlx/ops.h"
#include "random/common.h"

namespace mlx_sparse {

namespace {

template <typename I>
std::tuple<mx::array, mx::array>
random_csc_indices_host(mx::array key, int64_t n_rows, int64_t n_cols,
                        int64_t nnz, mx::Dtype index_dtype) {
  key.eval();
  std::vector<uint64_t> column_major(static_cast<size_t>(nnz));
  if (nnz > 0) {
    const auto *key_ptr = key.data<uint32_t>();
    const uint64_t seed = keyed_seed(key_ptr, n_rows, n_cols, nnz);
    const uint64_t total =
        static_cast<uint64_t>(n_rows) * static_cast<uint64_t>(n_cols);
    const uint64_t n_cols_u = static_cast<uint64_t>(n_cols);
    const uint64_t n_rows_u = static_cast<uint64_t>(n_rows);
    const int workers = configured_cpu_worker_count();
    auto fill_range = [&](CpuRange range) {
      for (int64_t i = range.begin; i < range.end; ++i) {
        const uint64_t linear =
            random_linear_index(static_cast<uint64_t>(i), total, seed);
        const uint64_t row = linear / n_cols_u;
        const uint64_t col = linear % n_cols_u;
        column_major[static_cast<size_t>(i)] = col * n_rows_u + row;
      }
    };
    if (workers > 1 && nnz > 0) {
      parallel_for_cpu_ranges(equal_cpu_ranges(static_cast<int>(nnz), workers),
                              fill_range);
    } else {
      fill_range({0, static_cast<int>(nnz)});
    }
    std::sort(column_major.begin(), column_major.end());
  }

  std::vector<I> indices(static_cast<size_t>(nnz));
  std::vector<I> indptr(static_cast<size_t>(n_cols) + 1, I{0});
  size_t read = 0;
  for (int64_t col = 0; col < n_cols; ++col) {
    while (read < column_major.size() &&
           column_major[read] / static_cast<uint64_t>(n_rows) ==
               static_cast<uint64_t>(col)) {
      indices[read] =
          static_cast<I>(column_major[read] % static_cast<uint64_t>(n_rows));
      ++read;
    }
    indptr[static_cast<size_t>(col) + 1] = static_cast<I>(read);
  }
  if (read != column_major.size()) {
    throw std::runtime_error("random_csc_indices internal count mismatch.");
  }

  return {
      mx::array(indices.begin(), mx::Shape{static_cast<int>(nnz)}, index_dtype),
      mx::array(indptr.begin(), mx::Shape{static_cast<int>(n_cols + 1)},
                index_dtype)};
}

} // namespace

std::tuple<mx::array, mx::array>
random_csc_indices(const mx::array &key, int64_t n_rows, int64_t n_cols,
                   int64_t nnz, int index_dtype_bits, mx::StreamOrDevice s) {
  check_random_key(key);
  const auto index_dtype = random_index_dtype_from_bits(index_dtype_bits);
  check_random_shape(n_rows, n_cols, nnz, index_dtype);

  auto stream = mx::to_stream(s);
  auto key_contig = mx::contiguous(key, false, stream);
  if (nnz == 0) {
    return {mx::zeros(mx::Shape{0}, index_dtype, stream),
            mx::zeros(mx::Shape{static_cast<int>(n_cols + 1)}, index_dtype,
                      stream)};
  }
  if (stream.device == mx::Device::cpu) {
    if (index_dtype == mx::int32) {
      return random_csc_indices_host<int32_t>(std::move(key_contig), n_rows,
                                              n_cols, nnz, index_dtype);
    }
    return random_csc_indices_host<int64_t>(std::move(key_contig), n_rows,
                                            n_cols, nnz, index_dtype);
  }

  auto counts = random_compressed_counts(key_contig, n_rows, n_cols, nnz,
                                         /*csc=*/true, stream);
  auto indptr = random_compressed_indptr(counts, index_dtype, stream);
  auto keys = random_structural_keys(key_contig, n_rows, n_cols, nnz,
                                     /*csc=*/true, stream);
  auto sorted_keys = mx::sort(keys, stream);
  auto indices = random_compressed_unpack_sorted_keys(
      sorted_keys, n_rows, index_dtype, /*csc=*/true, stream);
  return {indices, indptr};
}

} // namespace mlx_sparse
